### ISSUE TITLE
The FA_wrapper is not working correctly

### ISSUE THREAD

**Issue by HaFred** [ChFrenkel/DirectRandomTargetProjection#1]
Hi @ChFrenkel , thank you for sharing such an amazing work. When I tried to reproduce your results, I found that, `forward` here always return simply `self.module(x)`, which makes the `FA` is working as `BP` as a matter of fact. https://github.com/ChFrenkel/DirectRandomTargetProjection/blob/b84e3fe965bf98b8531f3036604cbe16424a5cc6/module.py#L61
A possible solution is to pass in the `requires_grad=True` for `FA`?

**Comment by ChFrenkel** [comment]
Hi @HaFred,
The purpose of the FA wrapper is only to change the gradients that are propagated to the layer that comes before it, based on the gradient that it received from the layer after it.
The wrapper does not need to do anything else: the weight updates are computed as for BP based on the gradient that was received from the layer after and on the locally-available activation function derivative and layer input.
The "if x.requires_grad:" condition is only there to make sure that you are both in the training phase (otherwise there are no gradients involved) and that the current layer is not the input layer (as there is no layer before it, so no gradient to propagate).
Does that answer your question?

**Comment by HaFred** [comment]
> Hi @HaFred,
> The purpose of the FA wrapper is only to change the gradients that are propagated to the layer that comes before it, based on the gradient that it received from the layer after it.

I agree.


> The wrapper does not need to do anything else: the weight updates are computed as for BP based on the gradient that was received from the layer after and on the locally-available activation function derivative and layer input.
> The "if x.requires_grad:" condition is only there to make sure that you are both in the training phase (otherwise there are no gradients involved) and that the current layer is not the input layer (as there is no layer before it, so no gradient to propagate).
> Does that answer your question?

Actually, what I am trying to say is that, based on my experiments, even though for the training process, the forward method cannot go through line55-59 because `requires_grad` is not explicitly passed in as `True`. Am I missing anything? Could you try to set a breakpoint in line 55 to see whether it is processed?
https://github.com/ChFrenkel/DirectRandomTargetProjection/blob/b84e3fe965bf98b8531f3036604cbe16424a5cc6/module.py#L55

**Comment by ChFrenkel** [comment]
I am really surprised that you encounter such an issue.
Starting from the code that is currently on GitHub with no other modification, launching with "python3 main.py --train-mode FA --topology FC_100_FC_100_FC_10", we get during the training phase that hooks are created for the output and hidden layers, not for the input one, which is the expected behavior.
We just verified this with two different installs based on PyTorch 1.3.0 and 1.4.0.
Could you try again with a clean PyTorch environment and the code without modifications?

**Comment by HaFred** [comment]
Thanks for the reply. You are right. I just realized that for the original codes, during the backward pass, only the first hidden layer will be with `requires_grad=False`. So in such a case, I think even without the `if x.requires_grad` condition, the logic still holds, right? (For the first layer, even though the error grad is calculated and useless, it does not affect anything; for the testing phase, the grads are redundant but still won't affect)

**Comment by ChFrenkel** [comment]
You are correct, the gradients that would be computed in that first layer would be useless.
However, if you remove this `if x.requires_grad` condition, an error will be raised for the command `x.register_hook(self.FA_hook_pre)`, as no hook can be attached to a variable that does not require gradient computation.
Adding this condition is thus the cleanest way to avoid this issue.
Does that answer your question?

**Comment by HaFred** [comment]
> You are correct, the gradients that would be computed in that first layer would be useless.
> However, if you remove this `if x.requires_grad` condition, an error will be raised for the command `x.register_hook(self.FA_hook_pre)`, as no hook can be attached to a variable that does not require gradient computation.
> Adding this condition is thus the cleanest way to avoid this issue.
> Does that answer your question?

Yes, that could not be more clear to me, thank you. I also tried to hardcode the `requires_grad` in the `train_epoch()` with `data.requires_grad_()` to avoid the error you mentioned, and it seems that this workaround could make the input layer require_grad, but it can only work for one epoch. I guess it is because in the `test_epoch()`, `FA_wrapper.forward` still needs the `x.requires_grad` condition. Thanks to the mechanism of autograd, this condition is really an elegant way to avoid the error.

**Comment by HaFred** [comment]
Hi @ChFrenkel, if you don't mind, could you help me with a simple doubt on the DRTP? So I noticed that the `grad_in` or in your code `grad_output_est` for the backward method of the TrainingHook is defined as `grad_output_est = labels.mm(fixed_fb_weights.view(-1,prod(fixed_fb_weights.shape[1:]))).view(grad_output.shape)` when it is for DRTP. Does not it mean that the gradient for a specific activation element is fixed?
In such case, I guess the weight_grad would be `weight_grad_(layer_i) = activation_(layer_i-1) * activation _grad_(layer_i)`, so in such case the `weight_grad` will be determined by activation only, rather than including activation_grad. Even though the DRTP as a quite aligned angle with BP, the learning capability seems degraded because the loss seems not involved just as FA?
What do you think? Thanks again.

**Comment by ChFrenkel** [comment]
Indeed, the gradient for a specific activation element **and for a specific label** is fixed for DRTP.
This is one of the key tradeoffs for DRTP: you can compute approximated layerwise loss gradients without any feedback (i.e. in a feedforward manner) and locally while still remaining aligned within 90° of the BP loss gradients (hence leading to cheap hardware implementations), but at the expense of an accuracy reduction.

**Comment by HaFred** [comment]
Great. Good to know.

**Comment by HaFred** [comment]
Great. Good to know.
